poseidon2/poseidon2_goldilocks.go

```diff
@@ -0,0 +1,163 @@
+package poseidon2
+
+import (
+	"github.com/Electron-Labs/plonky2-groth16-verifier/goldilocks"
+	"github.com/consensys/gnark/frontend"
+)
+
+type Poseidon2Goldilocks struct{}
+
+func MatmulM4(api frontend.API, rangeChecker frontend.Rangechecker, inputs []goldilocks.GoldilocksVariable) {
+	const t4 = WIDTH / 4
+	for i := 0; i < t4; i++ {
+		start_index := i * 4
+		t_0 := api.Add(inputs[start_index].Limb, inputs[start_index+1].Limb)
+		t_1 := api.Add(inputs[start_index+2].Limb, inputs[start_index+3].Limb)
+		t_2 := api.Mul(t_1, 1)
+
+		t_2Copy := api.Mul(t_2, 1)
+		t_2 = api.MulAcc(t_2Copy, inputs[start_index+1].Limb, 2)
+
+		t_3 := api.Mul(t_0, 1)
+
+		t_3Copy := api.Mul(t_3, 1)
+		t_3 = api.MulAcc(t_3Copy, inputs[start_index+3].Limb, 2)
+
+		t_4 := api.Mul(t_3, 1)
+
+		t_4Copy := api.Mul(t_4, 1)
+		t_4 = api.MulAcc(t_4Copy, t_1, 4)
+
+		t_5 := api.Mul(t_2, 1)
+
+		t_5Copy := api.Mul(t_5, 1)
+		t_5 = api.MulAcc(t_5Copy, t_0, 4)
+
+		// t2 -> 66
+		// t3 -> 66
+		// t4 -> 67
+		// t5 -> 67
+		inputs[start_index] = goldilocks.Reduce(api, rangeChecker, api.Add(t_3, t_5), 68)
+		inputs[start_index+1] = goldilocks.Reduce(api, rangeChecker, t_5, 67)
+		inputs[start_index+2] = goldilocks.Reduce(api, rangeChecker, api.Add(t_2, t_4), 68)
+		inputs[start_index+3] = goldilocks.Reduce(api, rangeChecker, t_4, 67)
+	}
+}
+
+func MatmulExternal(api frontend.API, rangeChecker frontend.Rangechecker, inputs []goldilocks.GoldilocksVariable) {
+	// Applying cheap 4x4 MDS matrix to each 4-element part of the state
+	MatmulM4(api, rangeChecker, inputs)
+
+	// Applying second cheap matrix for t > 4
+	// Compute store = [M4, M4, M4] * x
+	t4 := WIDTH / 4
+	storedNoReduce := make([]frontend.Variable, 4)
+	stored := make([]goldilocks.GoldilocksVariable, 4)
+
+	for l := 0; l < 4; l++ {
+		storedNoReduce[l] = inputs[l].Limb
+		for j := 1; j < t4; j++ {
+			storedNoReduce[l] = api.Add(storedNoReduce[l], inputs[4*j+l].Limb)
+		}
+	}
+	for l := 0; l < 4; l++ {
+		stored[l] = goldilocks.Reduce(api, rangeChecker, storedNoReduce[l], 65)
```

Comment:
We can move this reduction after to the last step and reduce with `66` rather than `65`. Saves ~1000 constraints.